### PR TITLE
[Intel MKL] Optimize combinedNMS performance

### PR DESCRIPTION
**Pull request by guizili0**
This PR optimize the CombinedNonMaxSuppression op CPU kernel from  single thread to multi-thread.

### CODE REVIEW

**Comment by arun0930** [comment]
0xF5BA8D9f16FEc5F97E06aF87f1025B001616CF3D

**Comment by arun0930** [comment]
0xF5BA8D9f16FEc5F97E06aF87f1025B001616CF3D

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and sorry for the delay!

**Comment by penpornk** [comment]
Also, could you please post the benchmark numbers comparison between before and after the PR? Thank you very much!

File: tensorflow/core/kernels/non_max_suppression_op.cc [hunk 1]
```diff
@@ -284,14 +284,179 @@ void DoNonMaxSuppressionOp(OpKernelContext* context, const Tensor& scores,
...
+    }
+  }
+
+  std::vector<int> selected;
+  std::vector<float> selected_boxes;
+  Candidate next_candidate;
+
+  std::sort(candidate_vector.begin(), candidate_vector.end(), cmp);
+  const Tensor const_boxes = boxes;
+  typename TTypes<float, 2>::ConstTensor boxes_data_1 =
```
[the remark this answers is not in the document]
Comment (guizili0) [reply tensorflow/core/kernels/non_max_suppression_op.cc]: @penpornk The new function has the parameter named boxes_data, so I renamed this one.

File: tensorflow/core/kernels/non_max_suppression_op.cc [hunk 2]
```diff
@@ -300,165 +465,71 @@ void BatchedNonMaxSuppressionOp(
...
+  int length = num_batches * num_classes;
+  // Input data boxes_data, scores_data
+  int input_bytes = length * num_boxes * 5;
+  int output_bytes = length * num_boxes * 5;
+  int compute_cycles = (Eigen::TensorOpCost::AddCost<int>() * 5 +
+                        Eigen::TensorOpCost::MulCost<int>() * 2 +
+                        Eigen::TensorOpCost::AddCost<float>() * 10 +
+                        Eigen::TensorOpCost::MulCost<float>() * 6 +
+                        Eigen::TensorOpCost::DivCost<float>()) *
+                       length;
```
[the remark this answers is not in the document]
Comment (guizili0) [reply tensorflow/core/kernels/non_max_suppression_op.cc]: @penpornk Thanks for you detail comments on this part. This part is really needed from my site. Thank you very much. I have update my code by your comments, please help to check. Thank you!

**Comment by guizili0** [comment]
@penpornk For the performance data, here is some data I got before. For the update performance data, due to the original single thread code take very log time, I will update when it finished.
| Batch size | Single thread | Multi-thread | ratio  |
|------------|---------------|--------------|--------|
| 1          | 10.8613       | 1.95         | 5.57x  |
| 28         | 336.59        | 31.05        | 10.84x |
| 32         | 383.72        | 35.56        | 10.79x |
| 64         | 770.59        | 63.68        | 12.10x |

File: tensorflow/core/kernels/non_max_suppression_op_benchmark_test.cc [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    testing::ItemsProcessed(iters* B);                                       \
+    test::Benchmark(#DEVICE, BM_CombinedNonMaxSuppression(B, BN, CN, Q))     \
+        .Run(iters);                                                         \
+  }                                                                          \
+  BENCHMARK(BM_CombinedNMS_##DEVICE##_##B##_##BN##_##CN##_##Q);
+
+BM_CombinedNonMaxSuppressionDev(cpu, 1, 1917, 90, 1);
+BM_CombinedNonMaxSuppressionDev(cpu, 28, 1917, 90, 1);
+BM_CombinedNonMaxSuppressionDev(cpu, 32, 1917, 90, 1);
+BM_CombinedNonMaxSuppressionDev(cpu, 64, 1917, 90, 1);
```
[the remark this answers is not in the document]
Comment (penpornk) [reply tensorflow/core/kernels/non_max_suppression_op_benchmark_test.cc]: The changes look great. Thank you!

File: tensorflow/core/kernels/non_max_suppression_op.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (penpornk) [reply tensorflow/core/kernels/non_max_suppression_op.cc]: Thank you for the clarification!

File: tensorflow/core/kernels/non_max_suppression_op.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (penpornk) [reply tensorflow/core/kernels/non_max_suppression_op.cc]: Thank you for the changes!

**Review by penpornk — APPROVED** [review]
Thank you for the fixes and for the performance data! How many cores was the benchmark run on?

**Comment by guizili0** [comment]
@penpornk update the performance data, time in ns
BM_CombinedNMS_cpu_1_500_25_1		2323039 ->	489212
BM_CombinedNMS_cpu_28_500_25_1		58247530 ->	9023750
BM_CombinedNMS_cpu_32_500_25_1		68019130 ->	9980150
BM_CombinedNMS_cpu_64_500_25_1		133137160 ->	17240980
BM_CombinedNMS_cpu_1_1000_25_1		2623302 ->	467080
BM_CombinedNMS_cpu_28_1000_25_1		71364650 ->	9396000
BM_CombinedNMS_cpu_32_1000_25_1		81346500 ->	10320460
BM_CombinedNMS_cpu_64_1000_25_1		164225940 ->	17787690
BM_CombinedNMS_cpu_1_1917_25_1		3652162 ->	538108
BM_CombinedNMS_cpu_28_1917_25_1		94482740 ->	9732510
BM_CombinedNMS_cpu_32_1917_25_1		108465480 ->	10609850
BM_CombinedNMS_cpu_64_1917_25_1		217078980 ->	18926130
BM_CombinedNMS_cpu_1_2500_25_1		4283823 ->	517773
BM_CombinedNMS_cpu_28_2500_25_1		111175170 ->	10150130
BM_CombinedNMS_cpu_32_2500_25_1		127008710 ->	11184190
BM_CombinedNMS_cpu_64_2500_25_1		253058940 ->	19643400
BM_CombinedNMS_cpu_1_500_25_25		2253545 ->	454969
BM_CombinedNMS_cpu_28_500_25_25		56916130 ->	9267910
BM_CombinedNMS_cpu_32_500_25_25		66019460 ->	10186830
BM_CombinedNMS_cpu_64_500_25_25		133661190 ->	17726410
BM_CombinedNMS_cpu_1_1000_25_25		2831089 ->	463359
BM_CombinedNMS_cpu_28_1000_25_25		70988320 ->	9753240
BM_CombinedNMS_cpu_32_1000_25_25		83533500 ->	10642970
BM_CombinedNMS_cpu_64_1000_25_25		167970350 ->	18761130
BM_CombinedNMS_cpu_1_1917_25_25		3525720 ->	521659
BM_CombinedNMS_cpu_28_1917_25_25		98398690 ->	10358520
BM_CombinedNMS_cpu_32_1917_25_25		112060740 ->	11647260
BM_CombinedNMS_cpu_64_1917_25_25		233556880 ->	20978420
BM_CombinedNMS_cpu_1_2500_25_25		4191191 ->	597319
BM_CombinedNMS_cpu_28_2500_25_25		113747490 ->	11175120
BM_CombinedNMS_cpu_32_2500_25_25		132507210 ->	12499690
BM_CombinedNMS_cpu_64_2500_25_25		267890750 ->	22600950
BM_CombinedNMS_cpu_1_500_90_1		7309910 ->	1847117
BM_CombinedNMS_cpu_28_500_90_1		200441650 ->	29992800
BM_CombinedNMS_cpu_32_500_90_1		233193640 ->	33024660
BM_CombinedNMS_cpu_64_500_90_1		470689830 ->	58343810
BM_CombinedNMS_cpu_1_1000_90_1		8930110 ->	1825090
BM_CombinedNMS_cpu_28_1000_90_1		246605820 ->	30728370
BM_CombinedNMS_cpu_32_1000_90_1		278993340 ->	34076850
BM_CombinedNMS_cpu_64_1000_90_1		568423760 ->	60270210
BM_CombinedNMS_cpu_1_1917_90_1		12124350 ->	1900205
BM_CombinedNMS_cpu_28_1917_90_1		338306180 ->	32426880
BM_CombinedNMS_cpu_32_1917_90_1		384421020 ->	35914810
BM_CombinedNMS_cpu_64_1917_90_1		774663220 ->	64348500
BM_CombinedNMS_cpu_1_2500_90_1		14097700 ->	1884557
BM_CombinedNMS_cpu_28_2500_90_1		396658660 ->	34139710
BM_CombinedNMS_cpu_32_2500_90_1		454546510 ->	37974240
BM_CombinedNMS_cpu_64_2500_90_1		909600120 ->	68007810
BM_CombinedNMS_cpu_1_500_90_90		7508650 ->	1848519
BM_CombinedNMS_cpu_28_500_90_90		208535130 ->	30469320
BM_CombinedNMS_cpu_32_500_90_90		240093220 ->	33806870
BM_CombinedNMS_cpu_64_500_90_90		480372590 ->	60929960
BM_CombinedNMS_cpu_1_1000_90_90		9130900 ->	1869414
BM_CombinedNMS_cpu_28_1000_90_90		267777270 ->	32220050
BM_CombinedNMS_cpu_32_1000_90_90		307633500 ->	35853710
BM_CombinedNMS_cpu_64_1000_90_90		626706220 ->	64038160
BM_CombinedNMS_cpu_1_1917_90_90		12437620 ->	1909809
BM_CombinedNMS_cpu_28_1917_90_90		383174840 ->	34798410
BM_CombinedNMS_cpu_32_1917_90_90		438880640 ->	38745970
BM_CombinedNMS_cpu_64_1917_90_90		889625800 ->	70248640
BM_CombinedNMS_cpu_1_2500_90_90		14539030 ->	1956012
BM_CombinedNMS_cpu_28_2500_90_90		457186240 ->	36712930
BM_CombinedNMS_cpu_32_2500_90_90		524073030 ->	41214980
BM_CombinedNMS_cpu_64_2500_90_90		1062754490 ->	74883870
BM_CombinedNMS_cpu_1_500_200_1		15750910 ->	3826558
BM_CombinedNMS_cpu_28_500_200_1		453637270 ->	55639170
BM_CombinedNMS_cpu_32_500_200_1		511463770 ->	62462100
BM_CombinedNMS_cpu_64_500_200_1		1026653820 ->	118736620
BM_CombinedNMS_cpu_1_1000_200_1		19583570 ->	3644278
BM_CombinedNMS_cpu_28_1000_200_1		547092330 ->	57528180
BM_CombinedNMS_cpu_32_1000_200_1		628781270 ->	64561770
BM_CombinedNMS_cpu_64_1000_200_1		1251330520 ->	121423990
BM_CombinedNMS_cpu_1_1917_200_1		26624460 ->	3749215
BM_CombinedNMS_cpu_28_1917_200_1		755978880 ->	61018550
BM_CombinedNMS_cpu_32_1917_200_1		871875440 ->	68690820
BM_CombinedNMS_cpu_64_1917_200_1		1747990780 ->	129082990
BM_CombinedNMS_cpu_1_2500_200_1		30810760 ->	3844000
BM_CombinedNMS_cpu_28_2500_200_1		886065080 ->	64371340
BM_CombinedNMS_cpu_32_2500_200_1		1012045410 ->	72465650
BM_CombinedNMS_cpu_64_2500_200_1		2065319100 ->	138367530
BM_CombinedNMS_cpu_1_500_200_200		16516840 ->	3667984
BM_CombinedNMS_cpu_28_500_200_200		468808630 ->	57981460
BM_CombinedNMS_cpu_32_500_200_200		530057360 ->	64955740
BM_CombinedNMS_cpu_64_500_200_200		1070955330 ->	122629750
BM_CombinedNMS_cpu_1_1000_200_200		20182400 ->	3741022
BM_CombinedNMS_cpu_28_1000_200_200		587699560 ->	60782560
BM_CombinedNMS_cpu_32_1000_200_200		671253770 ->	68633970
BM_CombinedNMS_cpu_64_1000_200_200		1348274040 ->	129640540
BM_CombinedNMS_cpu_1_1917_200_200		27440800 ->	3876423
BM_CombinedNMS_cpu_28_1917_200_200		827276050 ->	67455700
BM_CombinedNMS_cpu_32_1917_200_200		941913330 ->	75362430
BM_CombinedNMS_cpu_64_1917_200_200		1934794720 ->	143265870
BM_CombinedNMS_cpu_1_2500_200_200		32190200 ->	3975077
BM_CombinedNMS_cpu_28_2500_200_200		983298270 ->	72819920
BM_CombinedNMS_cpu_32_2500_200_200		1126059520 ->	81536040
BM_CombinedNMS_cpu_64_2500_200_200		2286195960 ->	154580820

**Comment by guizili0** [comment]
@penpornk this is a 2 socket system with 28 cores each. Thank you!

**Comment by penpornk** [comment]
@guizili0 Thank you very much! :)

File: tensorflow/core/kernels/non_max_suppression_op.cc [hunk 4]
```diff
@@ -284,14 +284,176 @@ void DoNonMaxSuppressionOp(OpKernelContext* context, const Tensor& scores,
...
   const int num_batches = inp_boxes.dim_size(0);
+  int num_classes = inp_scores.dim_size(2);
+  int q = inp_boxes.dim_size(2);
+
+  const float* scores_data =
+      const_cast<float*>(inp_scores.flat<float>().data());
+  const float* boxes_data = const_cast<float*>(inp_boxes.flat<float>().data());
+
+  int boxes_per_batch = num_boxes * q * 4;
+  int scores_per_batch = num_boxes * num_classes;
```
Comment (penpornk) [inline tensorflow/core/kernels/non_max_suppression_op.cc]: These gave unused variables errors. Please capture them in `shard_nms` and delete lines 468-469. Thank you very much!

**Review by penpornk — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/non_max_suppression_op.cc (hunk 4 above)
Comment (guizili0) [reply tensorflow/core/kernels/non_max_suppression_op.cc]: @penpornk Done, thanks.

**Review by penpornk — APPROVED** [review]
Thank you again!

**Pull request merged**